File: 0503-next-greater-element-ii/0503-next-greater-element-ii.cpp
class Solution {
public:
    vector<int> nextGreaterElements(vector<int>& nums) {
        stack<int> st;
    int n = nums.size();
    // vector<int> arr=nums;
    // for(int i=0;i<n;i++)
    //     arr.push_back(nums[i]);
    
    vector<int> ans(n, -1);
    for (int i = 2*n - 1; i >= 0; i--)
    {
        while (!st.empty() && st.top() <= nums[i%n])
            st.pop();
        if (st.empty())
            ans[i%n] = -1;
        else
            ans[i%n] = st.top();
        st.push(nums[i%n]);
    }
    // vector<int> ans2(ans.begin(),ans.begin()+n);
    // return ans2;
    return ans;
    }
};

File: 0503-next-greater-element-ii/0503-next-greater-element-ii.java
class Solution {
    public int[] nextGreaterElements(int[] nums) {
        List<Integer> arr = new ArrayList<>();
        int n = nums.length * 2;
        for (int num : nums) {
            arr.add(num);
        }
        for (int num : nums) {
            arr.add(num);
        }
        Stack<Integer> st = new Stack<>();
        int[] ans = new int[n];
        ans[n - 1] = -1;
        st.add(arr.get(n - 1));
        for (int i = n - 2; i >= 0; i--) {
            while (!st.isEmpty() && st.peek() <= arr.get(i)) {
                st.pop();
            }
            if (st.isEmpty()) {
                ans[i] = -1;
            } else {
                ans[i] = st.peek();
            }
            st.add(arr.get(i));
        }
        return Arrays.copyOfRange(ans, 0, n/2);
    }
}